### PR TITLE
Correct handling of renewed_to

### PR DESCRIPTION
**Pull request by chris124567**
This implements the first part of https://github.com/SiaFoundation/indexd/issues/227... I wanted to make sure I was on the right track before continuing with the pruning test since I'm still somewhat inexperienced with indexd.

### CODE REVIEW

File: contracts/chain_test.go [hunk 1]
```diff
@@ -340,7 +343,7 @@ func (tx *mockUpdateTx) ContractElements() ([]types.V2FileContractElement, error
 	return stateElements, nil
 }
 
-func (tx *mockUpdateTx) Contract(contractID types.FileContractID) (types.V2FileContractElement, ContractState) {
+func (tx *mockUpdateTx) Contract(contractID types.FileContractID) (types.V2FileContractElement, ContractState, *types.FileContractID) {
```
Comment (peterjan) [inline contracts/chain_test.go]: is updating the signature here a leftover? 

edit: I missed the test, it's a bit of a `nit` but I would remove this and add an extra method called `Renewal` or `RenewedContract`, just because the signature is a little weird and it's not obvious what you're returning.

File: persist/postgres/contracts.go [hunk 2]
```diff
@@ -497,6 +497,19 @@ func (tx *updateTx) UpdateContractState(contractID types.FileContractID, state c
 	return nil
 }
 
+// UpdateContractRenewedTo updates the renewed to ID of a contract to the provided one.
+func (tx *updateTx) UpdateContractRenewedTo(contractID types.FileContractID, renewedTo *types.FileContractID) error {
+	var value any
+	if renewedTo != nil {
+		value = sqlHash256(*renewedTo)
+	}
+	_, err := tx.tx.Exec(tx.ctx, `UPDATE contracts SET renewed_to = $1 WHERE contract_id = $2`, value, sqlHash256(contractID))
```
Comment (peterjan) [inline persist/postgres/contracts.go]: I think it makes sense to make this query a bit more strict, we really want this to error out if this is not doing exactly what we expect it's doing. E.g. we want to allow going from `renewed_to` -> `NULL` and from `NULL` to `renewed_to`, we don't want to allow `NULL` -> `NULL` (fine of course, but not expected to happen) or `renwed_to_a` -> `renewed_to_b` (not good, we should've nulled it before the update). 

I would split it up depending on whether `renewedTo` is nil, then do `WHERE contract_id = $2 AND renwed_to IS NULL` and use `res.RowsAffected` to ensure we've updated a row.

**Comment by peterjan** [comment]
On the right track 👍

**Comment by chris124567** [comment]
There are a few ways to proceed here but I'm not confident in any of them.  One easy way to do the pruning part would be to adapt the existing `TestPruneExpiredContractElements` in the postgres store,  Unfortunately, the `applyContractDiff`/`revertContractDiff` methods on the contract manager are private and it would seem inconsistent with the other managers to make those public.  There's no way to create a consensus.ApplyUpdate with arbitrary diffs as those fields are private too.  Alternatively, we could mine/revert blocks with renewals and create a sort of local test node like we have in the explored tests.  Or maybe we could mock a bunch of things in `contracts/chain_test.go`, but that wouldn't test the postgres store.

File: persist/postgres/contracts.go [hunk 3]
```diff
@@ -497,6 +497,21 @@ func (tx *updateTx) UpdateContractState(contractID types.FileContractID, state c
 	return nil
 }
 
+// UpdateContractRenewedTo updates the renewed to ID of a contract to the provided one.
+func (tx *updateTx) UpdateContractRenewedTo(contractID types.FileContractID, renewedTo *types.FileContractID) error {
```
Comment (ChrisSchinnerl) [inline persist/postgres/contracts.go]: This is missing a test

**Comment by ChrisSchinnerl** [comment]
> There are a few ways to proceed here but I'm not confident in any of them. One easy way to do the pruning part would be to adapt the existing `TestPruneExpiredContractElements` in the postgres store, Unfortunately, the `applyContractDiff`/`revertContractDiff` methods on the contract manager are private and it would seem inconsistent with the other managers to make those public. There's no way to create a consensus.ApplyUpdate with arbitrary diffs as those fields are private too. Alternatively, we could mine/revert blocks with renewals and create a sort of local test node like we have in the explored tests. Or maybe we could mock a bunch of things in `contracts/chain_test.go`, but that wouldn't test the postgres store.

I think the second test will need to be in `persist/postgres/contracts_test.go` like `TestPruneExpiredContractElements` but instead of using `applyContractDiff`/`revertContractDiff` you fake the revert by manually updating the contracts.
Meaning you set the `state` of each contract and the `renewed_to` field of the parent manually. Then you add the new renewed contract followed by rejecting and pruning expired contracts.

**Review by ChrisSchinnerl — CHANGES_REQUESTED** [review]

File: persist/postgres/contracts.go [hunk 4]
```diff
@@ -498,6 +498,21 @@ func (s *Store) RejectPendingContracts(ctx context.Context, maxFormation time.Ti
 	})
 }
 
+// UpdateContractRenewedTo updates the renewed to ID of a contract to the provided one.
+func (tx *updateTx) UpdateContractRenewedTo(contractID types.FileContractID, renewedTo *types.FileContractID) error {
+	var value any
+	if renewedTo != nil {
+		value = sqlHash256(*renewedTo)
+	}
+	res, err := tx.tx.Exec(tx.ctx, `UPDATE contracts SET renewed_to = $1 WHERE contract_id = $2`, value, sqlHash256(contractID))
```
Comment (ChrisSchinnerl) [inline persist/postgres/contracts.go]: You should be able to drop the if condition if you do the following

```suggestion
	res, err := tx.tx.Exec(tx.ctx, `UPDATE contracts SET renewed_to = $1 WHERE contract_id = $2`, (*sqlHash256)(renewedTo), sqlHash256(contractID))
```

File: contracts/chain.go [hunk 5]
```diff
@@ -123,6 +124,12 @@ func (m *ContractManager) applyContractDiff(tx *updateTx, diff consensus.V2FileC
 		switch {
 		case diff.Resolution != nil:
 			state = ContractStateResolved
+			if _, ok := diff.Resolution.(*types.V2FileContractRenewal); ok {
+				renewedTo := diff.V2FileContractElement.ID.V2RenewalID()
+				if err := tx.UpdateContractRenewedTo(diff.V2FileContractElement.ID, &renewedTo); err != nil {
+					return fmt.Errorf("failed to update renewed to for %v: %w", diff.V2FileContractElement.ID, err)
+				}
+			}
```
Comment (ChrisSchinnerl) [inline contracts/chain.go:132]: I think we might be running into a potential panic here when somehow the contract that we renewed to gets deleted. Because `renewed_to` has a foreign key constraint on `contracts.contract_id`. So we need to change `UpdateContractRenewedTo` to not update `renewed_to` if the contract doesn't exist and explain why that is in the docstring. And add that case to `TestUpdateContractRenewedTo`.

File: persist/postgres/contracts_test.go [hunk 6]
```diff
@@ -1270,6 +1270,199 @@ func TestUpdateContractRevision(t *testing.T) {
...
+	}
+
+	if contract, err := store.Contract(context.Background(), contractID); err != nil {
+		t.Fatal(err)
+	} else if contract.RenewedTo != (types.FileContractID{}) {
+		t.Fatal("expected contract to not be renewed")
+	}
+}
+
+func TestPruneRenewedTo(t *testing.T) {
```
Comment (ChrisSchinnerl) [inline persist/postgres/contracts_test.go]: While looking at this I realized that what we do in this test can't happen in RHP4 making this test obsolete. Because a contract can't renew to 2 different IDs in RHP4 anymore. Meaning that the other tests in this PR should already cover everything we need.

However, this reminded me of another edge case that I opened an issue for: https://github.com/SiaFoundation/indexd/issues/347

**Review by ChrisSchinnerl — CHANGES_REQUESTED** [review]

File: contracts/chain.go (hunk 5 above)
Comment (chris124567) [reply contracts/chain.go:132]: Do we want to return nil in that case?  The query as it is now would error if executing it would cause a constraint to be violated.

**Review by ChrisSchinnerl — APPROVED** [review]

**Review by n8mgr — APPROVED** [review]
